src/lib.rs

```
@@ -289,11 +290,218 @@ impl<I2C: embedded_hal_async::i2c::I2c> Lis2dw12<I2C> {
+
+        // Control 2: Enable Block Data Update
+        let control2: u8 = self
+            .modify_reg_field(
+                Register::Control2,
+                ControlReg2::new(
+                    false, false, false, true, // Block Data Update:
+                    false, Res0, false, false,
+                )
+                .into(),
+                0b1000,
+            )
+            .await?;
+
+        // Control 3: Single Data Conversion disabled
+        let control3: u8 = self
+            .modify_reg_field(
+                Register::Control3,
+                ControlReg3::new(
+                    false, // Single data conversion mode disabled
+                    false, // Single data conversion mode enabled by INT2 external trigger
+                    Res0,
+                    false,
+                    false,
+                    false,
+                    registers::Control3SelfTest::NormalMode, // Enabled after all controls set
+                )
+                .into(),
+                0b11, // Single data conversion mode bits
+            )
+            .await?;
+
+        // Control 6: Low noise Config Disabled, Full Scale Range 4g
+        let control6: u8 = self
+            .modify_reg_field(
+                Register::Control6,
+                ControlReg6::new(
+                    [Res0, Res0],
+                    false, // Disable low noise
+                    false,
+                    Control6FullScale::Scale4g,
+                    Control6BandwidthSelection::ODRdiv2,
+                )
+                .into(),
+                0b00110100, // Full Scale + Low Noise
+            )
+            .await?;
+
+        // 2. Record unbiased accelerometer samples
+        embassy_time::Timer::after_millis(TEST_STAGE_SLEEP_MS as u64).await;
+        self.flush_samples().await?;
+
+        let mut sample: usize = 0;
+        let mut attempts: usize = 0;
+        while sample < TEST_SAMPLES && attempts < MAX_SAMPLE_ATTEMPTS {
```

non_blocking: This is effectively a busy polling loop. Yielding asynchronously only during i2c transactions. How long does it take for the sensor to return data? Should we delay at least the sample period here? If something goes wrong, then we will be looping forever. Are we assuming that the client will call this and check for timeouts?

---

There was already the MAX_SAMPLE_ATTEMPTS variable to account for that so it won't go infinite. Added a delay of the sampling period after each sample and lowered the max attempts to reduce amount of polling

---

Doh! I saw `MAX_SAMPLE_ATTEMPTS` but my mind did not connect it. So in the timeout case, we would return zeros and fails the self test. 

Non-blocking, but for better traceability we should add actually return an error like https://github.com/OpenDevicePartnership/bq25773/blob/d718e329c8bad32f02154cd6bd618bd97255f6d9/src/lib.rs#L20C1-L21C1. This can be addressed after this PR and creation of an issue.

---

Yep! Or None to be precise for matching in the caller.
Sure, we can look into that afterwards, thanks.

---

Instead of returning an option we should return an error type defined by this driver:
`Result<(f32, f32, f32), Driver::Error>`. Having a result already present an invalid case of returning an error instead of a valid value. Adding an option on top of that just seems unnecessary

We can map any internal i2c error to a `Driver::Error` and just use the ? operator to pass the error up.

Please create an issue to follow up on this, thanks!

---

Yes that makes sense, and I think I get the pattern with the I2C error being a type within the Enum. Sounds good, issue created: https://github.com/OpenDevicePartnership/lis2dw12-i2c/issues/5